src/driver-sinowealth-nubwo.c

```diff
@@ -0,0 +1,315 @@
+	if (size < 0)
+		log_error(device->ratbag, "Error while sending pre-firmware request message: %d\n", size);
+
+	size = ratbag_hidraw_get_feature_report(device, SINOWEALTHNUBWO_GET_FIRMWARE_CMD_REPORTID, buff, SINOWEALTHNUBWO_GET_FIRMWARE_MSGSIZE);
+	if (size != SINOWEALTHNUBWO_GET_FIRMWARE_MSGSIZE)
+		log_error(device->ratbag ,"Firmware report reply size mismatch expected %d got %d\n", SINOWEALTHNUBWO_GET_FIRMWARE_MSGSIZE, size);
+	
+	char *fw_str = zalloc(SINOWEALTHNUBWO_GET_FIRMWARE_MSGSIZE); //Purposefully overallocate zeros to prevent buffer overrun.
+	memcpy(fw_str, buff + SINOWEALTHNUBWO_GET_FIRMWARE_MSGOFFSET, SINOWEALTHNUBWO_GET_FIRMWARE_MSGSIZE - SINOWEALTHNUBWO_GET_FIRMWARE_MSGOFFSET);
+	return fw_str;
+}
+
+static int
+sinowealthnubwo_probe(struct ratbag_device *device)
+{
+	int error;
+	struct ratbag_profile *profile;
+	struct ratbag_resolution *resolution;
+	struct ratbag_led *led;
+	
+	error = ratbag_find_hidraw(device, sinowealthnubwo_test_hidraw);
+	if (error)
+		return error;
+
+	ratbag_device_init_profiles(device,
+			SINOWEALTHNUBWO_NUM_PROFILES,
+			SINOWEALTHNUBWO_NUM_RESOLUTIONS,
+			SINOWEALTHNUBWO_NUM_BUTTONS,
+			SINOWEALTHNUBWO_NUM_LEDS);
+
+	ratbag_device_for_each_profile(device, profile) {
+		profile->is_active = true;
+
+		ratbag_profile_set_cap(profile, RATBAG_PROFILE_CAP_WRITE_ONLY);
+		ratbag_profile_set_report_rate_list(profile, REPORT_RATES, ARRAY_LENGTH(REPORT_RATES));
+		ratbag_profile_for_each_resolution(profile, resolution) {
+			ratbag_resolution_set_dpi_list(resolution, DPILIST, ARRAY_LENGTH(DPILIST));
+			resolution->dpi_x = resolution->dpi_y = DPILIST[0];
+			resolution->is_active = true;
+			resolution->is_default = true;
+		}
+		ratbag_profile_for_each_led(profile, led) {
+			led->mode = RATBAG_LED_OFF;
+			led->color.red = led->color.green = led->color.blue = 0;
+			led->colordepth = RATBAG_LED_COLORDEPTH_RGB_888;
+			led->type = RATBAG_LED_TYPE_SIDE;
+			led->color.red = led->color.green = led->color.blue = 0;
+			ratbag_led_set_mode_capability(led, RATBAG_LED_OFF);
+			ratbag_led_set_mode_capability(led, RATBAG_LED_ON);
+			ratbag_led_set_mode_capability(led, RATBAG_LED_BREATHING);
+			ratbag_led_set_mode_capability(led, RATBAG_LED_CYCLE);
+			//Actually more
+		}
+	}
+
+	char *fwstr = sinowealth_get_firmware_string(device);
+	log_info(device->ratbag, "Firmware: %s\n", fwstr);
+	free(fwstr);
+	return 0;
+}
+
+static uint8_t 
+encode_dpi(unsigned int dpi)
+{
+	for (size_t i = 0; i < ARRAY_LENGTH(DPILIST); i++)
+		if (DPILIST[i] == dpi)
+			return DPI_ENCODED[i];
+	return DPI_ENCODED[0];
+}
+
+static uint8_t 
+encode_report_rate(unsigned int reportrate)
+{
+	for (size_t i = 0; i < ARRAY_LENGTH(REPORT_RATES); i++)
+		if (REPORT_RATES[i] == reportrate)
+			return REPORT_RATES_ENCODED[i];
+	return REPORT_RATES_ENCODED[0];
+}
+
+static int
+sinowealthnubwo_set_dpi(struct ratbag_device *device, int dpi)
+{
+	uint8_t buf[SINOWEALTHNUBWO_PERF_CMD_MSGSIZE];
+	memset(buf, 0, SINOWEALTHNUBWO_PERF_CMD_MSGSIZE);
```

Comment:
You can also write this as `uint8_t buf[...] = {0};` :)

Reply:
TIL.